networkit/cpp/matching/SuitorMatcher.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ *  SuitorMatcher.cpp
+ *
+ *  Created on: 27.08.2019
+ *  Author: Michal Boron     <michal.s.boron@gmail.com>
+ *          Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+// networkit-format
+
+#include <algorithm>
+#include <stdexcept>
+
+#include <networkit/matching/SuitorMatcher.hpp>
+
+namespace NetworKit {
+
+SuitorMatcher::SuitorMatcher(const Graph &G, bool sortSuitor, bool checkSortedEdges)
+    : Matcher(G), sortSuitor(sortSuitor) {
+
+    if (G.numberOfSelfLoops() > 0)
+        throw std::runtime_error("This algorithm does not graphs with self-loops.");
+
+    if (G.isDirected())
+        throw std::runtime_error("This algorithm does not support directed graphs.");
+
+    if (sortSuitor && checkSortedEdges && !hasEdgesSortedByWeight(G))
+        throw std::runtime_error("Edges are not sorted by weight");
+}
+
+bool SuitorMatcher::hasEdgesSortedByWeight(const Graph &G) {
+    for (node u : G.nodeRange()) {
+        node prevNode = none;
+        edgeweight prevWeight = std::numeric_limits<edgeweight>::max();
+        bool sorted = true;
+        G.forNeighborsOf(u, [&](node v, edgeweight ew) {
+            if (ew == prevWeight) {
+                if (prevNode > v)
+                    sorted = false;
+
+            } else if (ew > prevWeight)
+                sorted = false;
+
+            prevNode = v;
+            prevWeight = ew;
+        });
+
+        if (!sorted)
+            return false;
+    }
+
+    return true;
+}
+
+void SuitorMatcher::findSuitor(node current) {
+    bool done = false;
+
+    do {
+
+        node partner = suitor[current];
+        edgeweight heaviest = ws[current];
+
+        G->forNeighborsOf(current, [&](const node v, const edgeweight weight) {
+            if ((weight > heaviest || (weight == heaviest && v < partner))
+                && (weight > ws[v] || (weight == ws[v] && current < suitor[v]))) {
+                partner = v;
+                heaviest = weight;
+            }
+        });
+
+        done = true;
+
+        if (partner != none
+            && (heaviest > ws[partner] || (heaviest == ws[partner] && current < suitor[partner]))) {
+            node y = suitor[partner];
+            suitor[partner] = current;
+            ws[partner] = heaviest;
+
+            // if the current vertex already has a suitor
+            if (y != none) {
+                current = y;
+                done = false;
+            }
+        }
+    } while (!done);
+}
+
+void SuitorMatcher::findSortSuitor(node current) {
+    bool done = false;
+
+    do {
+        node partner = suitor[current];
+        edgeweight heaviest = ws[current];
+
+        for (index &iter = neighborIterators[current]; iter < G->degree(current); ++iter) {
+            const node v = G->getIthNeighbor(current, iter);
+            const edgeweight weight = G->getIthNeighborWeight(current, iter);
```

Comment:
Why does `neighborIterators` not actually store an iterator (e.g., a `NeighborWeightRange`)?

Reply:
If the input graph is unweighted then `NeighborWeightRange` cannot be used because it requires an unweighted graph. To make this work we need to use two vectors with two different iterators and add some checks to pick the correct one at runtime, but I find it easier to just store the index to the current neighbor.
However, I think it would be a better idea to make `NeighborWeightRange` also work with unweighted graphs, that would make things easier.

Reply:
You mean "...because it requires a weighted graph", right?

Reply:
Yes, thanks.